internal/lint/publish.go

```diff
@@ -20,72 +20,123 @@ type PublisherInput struct {
+// It does publish and then subscribe.
+func PublisherPositionOldest(input *PublisherInput) {
+	linter := publisherPositionOldestLinter{
+		publisherLinter: &publisherLinter{
+			ctx:   context.Background(),
+			input: input,
+		},
+	}
+
+	ginkgo.BeforeEach(linter.Prepare)
+	ginkgo.AfterEach(linter.Cleanup)
+
+	ginkgo.Context("StartAt=PositionOldest", func() {
+		ginkgo.It("should publish", linter.Lint)
+	})
+}
 
-		cycle := time.Now().UnixNano()
-		topicA = fmt.Sprintf("bps-unittest-topic-%d-a", cycle)
-		topicB = fmt.Sprintf("bps-unittest-topic-%d-b", cycle)
+// ----------------------------------------------------------------------------
 
-		if input.Setup != nil {
-			Ω.Expect(input.Setup(topicA, topicB)).To(Ω.Succeed())
-		}
-	})
+type publisherLinter struct {
+	ctx            context.Context
+	input          *PublisherInput
+	topicA, topicB string
+}
 
-	ginkgo.AfterEach(func() {
-		if input.Teardown != nil {
-			Ω.Expect(input.Teardown(topicA, topicB)).To(Ω.Succeed())
-		}
-	})
+func (l *publisherLinter) Prepare() {
+	cycle := time.Now().UnixNano()
+	l.topicA = fmt.Sprintf("bps-unittest-topic-%d-a", cycle)
+	l.topicB = fmt.Sprintf("bps-unittest-topic-%d-b", cycle)
 
-	ginkgo.It("should publish", func() {
-		// run publishing and consuming concurrently (mostly for StartPosition=PositionNewest case):
-		var threads sync.WaitGroup
-
-		// produce concurrently:
-		threads.Add(1)
-		go func() {
-			defer threads.Done()
-			defer ginkgo.GinkgoRecover()
-
-			// give subscribers some time to actually start consuming (subscribe):
-			time.Sleep(subscriptionWaitDelay)
-
-			Ω.Expect(subject.Topic(topicA).Publish(ctx, &bps.PubMessage{Data: []byte("v1")})).To(Ω.Succeed())
-			Ω.Expect(subject.Topic(topicB).Publish(ctx, &bps.PubMessage{Data: []byte("v2")})).To(Ω.Succeed())
-			Ω.Expect(subject.Topic(topicA).Publish(ctx, &bps.PubMessage{Data: []byte("v3")})).To(Ω.Succeed())
-		}()
-
-		// consume different topics concurrently:
-		threads.Add(1)
-		go func() {
-			defer threads.Done()
-			defer ginkgo.GinkgoRecover()
-
-			Ω.Eventually(func() ([]*bps.PubMessage, error) {
-				return input.Messages(topicA, 2)
-			}, 3*subscriptionWaitDelay).Should(haveData("v1", "v3"))
-		}()
-
-		// consume different topics concurrently:
-		threads.Add(1)
-		go func() {
-			defer threads.Done()
-			defer ginkgo.GinkgoRecover()
-
-			Ω.Eventually(func() ([]*bps.PubMessage, error) {
-				return input.Messages(topicB, 1)
-			}, 3*subscriptionWaitDelay).Should(haveData("v2"))
-		}()
-
-		threads.Wait()
-	})
+	if l.input.Setup != nil {
+		Ω.Expect(l.input.Setup(l.topicA, l.topicB)).To(Ω.Succeed())
+	}
+}
+
+func (l *publisherLinter) Cleanup() {
+	if l.input.Teardown != nil {
+		Ω.Expect(l.input.Teardown(l.topicA, l.topicB)).To(Ω.Succeed())
+	}
+}
+
+// ----------------------------------------------------------------------------
+
+type publisherPositionNewestLinter struct {
+	*publisherLinter
+}
+
+func (l *publisherPositionNewestLinter) Lint() {
```

Comment:
Lint methods could probably be just inlined into exported funcs like PublisherPositionNewest(...)